dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpCdiExtension.java

```diff
@@ -69,14 +79,61 @@ void registerTools(@Observes ProcessManagedBean<?> pmb, BeanManager beanManager)
         }
     }
 
+    void discoverEncoderBeans(@Observes ProcessManagedBean<?> processManagedBean) {
+        AnnotatedType<?> type = processManagedBean.getAnnotatedBeanClass();
+        Class<?> javaClass = type.getJavaClass();
+        if (Encoder.class.isAssignableFrom(javaClass)) {
+            encoderBeans.add(processManagedBean.getBean());
+        } else if (EncoderRegistry.class.isAssignableFrom(javaClass)) {
+            encoderRegistryBean = processManagedBean.getBean();
+        }
+    }
+
     void afterDeploymentValidation(@Observes AfterDeploymentValidation afterDeploymentValidation, BeanManager manager) {
+        registerEncoders(manager);
+
         boolean error = reportOnDuplicateTools(afterDeploymentValidation) | reportOnToolArgEdgeCases(afterDeploymentValidation) |
                         reportOnDuplicateSpecialArguments(afterDeploymentValidation) | reportOnInvalidSpecialArguments(afterDeploymentValidation);
         if (error) {
             afterDeploymentValidation.addDeploymentProblem(new Exception(Tr.formatMessage(tc, "CWMCM0005E.validation.error")));
         }
     }
 
+    void registerEncoders(BeanManager beanManager) {
+        CreationalContext<?> context = beanManager.createCreationalContext(null);
+
+        encoderRegistry = (EncoderRegistry) beanManager.getReference(encoderRegistryBean, encoderRegistryBean.getBeanClass(), context);
```

Comment:
```suggestion
        encoderRegistry = beanManager.createInstance().select(EncoderRegistry.class);
```

We can do this in a slightly simpler way, without needing to store the `Bean` earlier, because it's our bean so our extension is able to find it.

Reply:
Done, thanks